formula/src/main/java/com/instacart/formula/FormulaContext.kt

```diff
@@ -2,120 +2,88 @@ package com.instacart.formula
 
 import com.instacart.formula.internal.JoinedKey
 import com.instacart.formula.internal.ScopedCallbacks
+import com.instacart.formula.internal.TransitionDispatcher
+import com.instacart.formula.internal.UnitCallback
 
 /**
  * Provides functionality within [evaluate][Formula.evaluate] function to [compose][child]
  * child formulas, handle events [FormulaContext.onEvent], and [respond][FormulaContext.updates]
  * to arbitrary asynchronous events.
  */
 abstract class FormulaContext<State> internal constructor(
-    @PublishedApi internal val callbacks: ScopedCallbacks
+    @PublishedApi internal val callbacks: ScopedCallbacks<State>,
+    internal val transitionDispatcher: TransitionDispatcher<State>,
 ) {
 
     /**
-     * Creates a callback to be used for handling UI event transitions.
-     *
-     * It uses inlined callback anonymous class for type.
-     */
-    inline fun onEvent(crossinline transition: Transition.Factory.() -> Transition<State>): () -> Unit {
-        val callback: (Unit) -> Unit = {
-            performTransition(transition(Transition.Factory))
-        }
-        val reference = callbacks.initOrFindCallback(callback::class)
-        reference.delegate = callback
-        return reference
-    }
-
-    /**
-     * Creates a callback to be used for handling UI event transitions.
-     *
-     * @param key key with which the callback is to be associated. Same key cannot be used for multiple callbacks.
-     */
-    inline fun onEvent(
-        key: Any,
-        crossinline transition: Transition.Factory.() -> Transition<State>
-    ): () -> Unit {
-        val callback = callbacks.initOrFindCallback(key)
-        callback.delegate = {
-            performTransition(transition(Transition.Factory))
-        }
-        return callback
-    }
-
-    /**
-     * Creates a callback that takes a [UIEvent] and performs a [Transition].
+     * Creates a callback that takes an [Event] and performs a [Transition].
      *
-     * It uses inlined callback anonymous class for type.
+     * It uses [transition] type as key.
      */
-    inline fun <UIEvent> onEvent(
```

Comment:
Any reason we migrated off `inline` for all functions here?

Reply:
Previously, we were creating another function inside of here and cross-inlining the `transition` function to save an extra object allocation. I've restructured this, so inline isn't needed anymore.